### PR TITLE
Added functionality to predict optimal multiplier ratio

### PR DESCRIPTION
**Pull request by RitwikC07**
#### Types of changes

- [ ] Bug fix (change which fixes an issue)
- [x] New feature (change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

#### Checklist:

- [x] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed

### CODE REVIEW

File: ODIN_II/SRC/MixingOptimization.cpp [hunk 1]
```diff
@@ -31,13 +31,16 @@
 #include "odin_error.h"           // error_message
 #include "adders.h"               // hard_adders
 #include "HardSoftLogicMixer.hpp" // HardSoftLogicMixer
+using namespace std;
```
Comment (sdamghan) [inline ODIN_II/SRC/MixingOptimization.cpp:34]: remove this line, where ever you need std namespace just use `std::`

File: ODIN_II/SRC/MixingOptimization.cpp [hunk 2]
```diff
@@ -31,13 +31,16 @@
...
 
-void MixingOpt::scale_counts() {
+void MixingOpt::scale_counts(float pred_ratio) {
     if (this->_blocks_count < 0 || this->_blocks_count == INT_MAX || this->_ratio < 0.0 || this->_ratio > 1.0) {
         error_message(NETLIST, unknown_location, "The parameters for optimization kind:%i are configured incorrectly : count %i, ratio %f\n", this->_kind, this->_blocks_count, this->_ratio);
         exit(0);
     }
+    this->_ratio = pred_ratio;
     this->_blocks_count = this->_blocks_count * this->_ratio;
+    std::cout<< "\n" << endl;
```
Comment (sdamghan) [inline ODIN_II/SRC/MixingOptimization.cpp:43]: what is this line? delete it

File: ODIN_II/SRC/memories.cpp [hunk 3]
```diff
@@ -35,6 +35,9 @@
 #include "partial_map.h"
 #include "vtr_util.h"
 #include "vtr_memory.h"
+#include <iostream>
+#include <fstream>
+using namespace std;
```
Comment (sdamghan) [inline ODIN_II/SRC/memories.cpp:40]: delete this line

File: ODIN_II/SRC/memories.cpp [hunk 4]
```diff
@@ -283,9 +286,11 @@ void add_output_port_to_memory(nnode_t* node, signal_list_t* signals, const char
  * Reports the memory distribution as well.
  */
 void check_memories_and_report_distribution() {
+    ofstream statsfile;
+    statsfile.open ("stats.txt");
     if ((sp_memory_list == NULL) && (dp_memory_list == NULL))
         return;
-
+
```
Comment (sdamghan) [inline ODIN_II/SRC/memories.cpp:293]: avoid having whitespace changes please

File: ODIN_II/SRC/memories.cpp [hunk 5]
```diff
@@ -35,6 +35,9 @@
 #include "partial_map.h"
 #include "vtr_util.h"
 #include "vtr_memory.h"
+#include <iostream>
```
Comment (sdamghan) [inline ODIN_II/SRC/memories.cpp:38]: I see you only used file stream, so in this case please avoid using extra includes like `iostream`

File: ODIN_II/SRC/odin_ii.cpp [hunk 6]
```diff
@@ -36,7 +41,6 @@
 
 #include "odin_globals.h"
 #include "odin_types.h"
-#include "odin_util.h"
```
Comment (sdamghan) [inline ODIN_II/SRC/odin_ii.cpp:39]: why did you remove odin_util?

File: ODIN_II/SRC/odin_ii.cpp [hunk 7]
```diff
@@ -68,6 +72,18 @@
 #include "GenericReader.hpp"
 #include "BLIF.hpp"
 
+
+#ifdef WINDOWS
```
Comment (sdamghan) [inline ODIN_II/SRC/odin_ii.cpp:76]: These macros are already added in odin_util, why didn't you use them? This project has a whole utility function, you would need to utilize them rather than writing everything from scratch

File: ODIN_II/SRC/odin_ii.cpp [hunk 8]
```diff
@@ -68,6 +72,18 @@
...
+#include <direct.h>
+#define GetCurrentDirectory _getcwd
+#else
+#include <unistd.h>
+#define GetCurrentDirectory getcwd
+#endif
+
+#include<iostream>
```
Comment (sdamghan) [inline ODIN_II/SRC/odin_ii.cpp:84]: delete it

File: ODIN_II/SRC/odin_ii.cpp [hunk 9]
```diff
@@ -68,6 +72,18 @@
...
+using namespace std;
```
Comment (sdamghan) [inline ODIN_II/SRC/odin_ii.cpp:85]: delete it

File: ODIN_II/SRC/odin_ii.cpp [hunk 10]
```diff
@@ -183,14 +199,50 @@ static void optimization() {
     printf("\n--------------------------------------------------------------------\n");
 }
 
+/*---------------------------------------------------------------------------
+ * (function: get_current_dir)
+ returns current working directory
+ *-------------------------------------------------------------------------*/
+std::string get_current_dir() {
```
Comment (sdamghan) [inline ODIN_II/SRC/odin_ii.cpp:206]: similar routine already exist in odin_util
get_directory
if you need additional feature try to add a new utility function there

File: ODIN_II/SRC/odin_ii.cpp [hunk 11]
```diff
@@ -183,14 +199,50 @@ static void optimization() {
...
+   GetCurrentDirectory( path_buffer, FILENAME_MAX );
+   string curr_working_dir(path_buffer);
+   return curr_working_dir;
+}
+
+/*---------------------------------------------------------------------------
+ * (function: predict_optimal_ratio)
+ returns optimal ratio for the current combination of circuit and architecture file
+ *-------------------------------------------------------------------------*/
+float predict_optimal_ratio(){
```
Comment (sdamghan) [inline ODIN_II/SRC/odin_ii.cpp:217]: This function must not be here in odin_ii.cpp
if you used mixing optimization, why didn't you add these predict optimal under the same subsystem?

File: vtr_flow/tasks/power/power_frac_sweep/config/config.txt [hunk 12]
```diff
@@ -13,22 +13,22 @@ circuits_dir=benchmarks/verilog
...
-circuit_list_add=bgm.v
-circuit_list_add=blob_merge.v
-circuit_list_add=boundtop.v
-circuit_list_add=ch_intrinsics.v
-circuit_list_add=diffeq1.v
+#circuit_list_add=bgm.v
+#circuit_list_add=blob_merge.v
+#circuit_list_add=boundtop.v
+#circuit_list_add=ch_intrinsics.v
+#circuit_list_add=diffeq1.v
```
Comment (sdamghan) [inline vtr_flow/tasks/power/power_frac_sweep/config/config.txt:20]: Remove all comments and anything modifications to this task. Again if you need, you would create a new one

File: ODIN_II/dt.py [hunk 13]
```diff
@@ -0,0 +1,28 @@
+import numpy as np
```
Comment (sdamghan) [inline ODIN_II/dt.py:1]: this file should be under the tools directory in the odin-ii/regression_tests
Also, you need to run `make format-py` and `./dev/pylint_check.py` from the VTR_ROOT directory

The same for C/C++ code -> `make format`

**Comment by poname** [comment]
Not finalized by participants.
